invdendpoint/customers.go

```diff
@@ -12,30 +12,36 @@ type Customers []Customer
-	PaymentSource      *PaymentSource         `json:"-"`                           //Customer’s payment source, if attached
-	PaymentSourceReady bool                   `json:"-"`                           //If true than run UnbundlePaymentSource()
-	Taxes              []Rate                 `json:"taxes,omitempty"`             //Collection of Tax Rate IDs
-	Type               string                 `json:"type,omitempty"`              //Organization type, company or person
-	AttentionTo        string                 `json:"attention_to,omitempty"`      //Used for ATTN: address line if company
-	Address1           string                 `json:"address1,omitempty"`          //First address line
-	Address2           string                 `json:"address2,omitempty"`          //Optional second address line
-	City               string                 `json:"city,omitempty"`              //City
-	State              string                 `json:"state,omitempty"`             //State or province
-	PostalCode         string                 `json:"postal_code,omitempty"`       //Zip or postal code
-	Country            string                 `json:"country,omitempty"`           //Two-letter ISO code
-	TaxID              string                 `json:"taxid,omitempty"`             //Tax ID to be displayed on documents
-	Phone              string                 `json:"phone,omitempty"`             //Phone #
-	Notes              string                 `json:"notes,omitempty"`             //Private customer notes
-	StatementPdfUrl    string                 `json:"statement_pdf_url,omitempty"` //URL to download the latest account statement
-	CreatedAt          int64                  `json:"created_at,omitempty"`        //Timestamp when created
-	MetaData           map[string]interface{} `json:"metadata,omitempty"`          //A hash of key/value pairs that can store additional information about this object.
+	SaveableCustomer
+	Id                 int64           `json:"id,omitempty"`                //The customer’s unique ID
+	CollectionMode     string          `json:"collection_mode,omitempty"`   //Invoice collection mode, auto or manual
+	PaymentSourceRAW   json.RawMessage `json:"payment_source,omitempty"`    //Holds the raw payment json information for later parsing
+	PaymentSource      *PaymentSource  `json:"-"`                           //Customer’s payment source, if attached
+	PaymentSourceReady bool            `json:"-"`                           //If true than run UnbundlePaymentSource()
+	StatementPdfUrl    string          `json:"statement_pdf_url,omitempty"` //URL to download the latest account statement
+	CreatedAt          int64           `json:"created_at,omitempty"`        //Timestamp when created
+}
+
+//SaveableCustomer includes the subset of Customer fields which are valid for customer creation/update requests
+type SaveableCustomer struct {
```

Comment:
🤔This struct, and really it's `StripeToken` field are making me think. If we have designs on eventually PR-ing our changes upstream and breaking free of this fork, I wonder if the name StripeToken for a string type is only tackling our use case or if Invoiced is a Stripe only shop so it doesn't matter.

Also, it's on me that CI isn't hooked up to this repo but I expect that revisions to this data structure would cause the `TestUnMarshalCustomerObject` test [here](https://github.com/ActiveState/invoiced-go/blob/5d407ab4f5456527c24428fec78b082eff514ecf/invdendpoint/customers_test.go#L9) to fail. Not a big deal either if we're not planning on bubbling this upstream though.

Reply:
`stripe_token` is a (write-only) `Customer` field defined in the Invoiced API itself: https://invoiced.com/docs/api/#create-a-customer. (This is distinct from the `invoiced_token` accepted by the calls for creating a payment source.) It's not that Invoiced is a Stripe-only shop, but that field does exist specifically for when Stripe is the payment gateway in use (and according to Invoiced via email, it _has_ to be used when Stripe is the gateway; Invoiced's own tokenization apparently won't work in that case).

Good call on running the tests! 😅 That lead me to a typo in the `stripe_customer`'s JSON struct tag, and I had to reorder fields in one test's marshalled JSON to account for the change to the Customer struct. `TestUnMarshalCustomerObject` was fine (the `stripe_customer` field being missing from the JSON isn't a problem, it just leaves the field empty in the resultant struct), but several other tests were failing on master, apparently as a result of https://github.com/ActiveState/invoiced-go/commit/da450eed1792d601bd8ea46e5d1ee9f07c9b0bef. I've updated them to expect `catalog_item` to be a string rather than an object, per that commit.

As for merging upstream, my intent is certainly for that to still be possible. The only opinionated thing I've done is the `SaveableCustomer` thing, which should be pretty much transparent, but certainly I could see Invoiced asking for a different approach at PR time.

Reply:
(Oh, and in case it's not clear, thanks to Go's syntax: the way struct composition works, the `Customer` struct effectively still has all the same fields it did before; you don't have to access the `SaveableCustomer` subset of fields via `customer.SaveableCustomer`, they can still be accessed and unmarshalled-into directly as `customer.foo` – hence my contention that it's a transparent and low-touch change.)

Reply:
😬 da450ee looks like my handy work, what sort of no-test-running foolishness was I getting up to!? Well, thanks for providing your thoughts/context. And for running the tests for this iteration 🙏